### PR TITLE
Make exit handling more robust

### PR DESCRIPTION
**Pull request by yyoncho**
- fixes 1160

### CODE REVIEW

File: lsp-mode.el [hunk 1]
```diff
@@ -5388,41 +5388,41 @@ returns the command to execute."
 (defun lsp--create-sentinel (workspace)
   "Create sentinel handler for WORKSPACE."
   (lambda (process exit-str)
-    (let* ((status (process-status process))
-           (folder->workspaces (lsp-session-folder->servers (lsp-session)))
-           (stderr (-> workspace lsp--workspace-proc process-name get-buffer)))
+    (unless (process-live-p process)
```
Comment (MaskRay) [inline lsp-mode.el]: Is there any early-return mechanism to reduce indentation below?

File: lsp-mode.el [hunk 2]
```diff
@@ -5388,41 +5388,41 @@ returns the command to execute."
...
+          ;; in the workspace.
+          (--each (lsp--workspace-buffers workspace)
+            (when (buffer-live-p it)
+              (with-current-buffer it
+                (setq lsp--buffer-workspaces (delete workspace lsp--buffer-workspaces))
+                (lsp--uninitialize-workspace)
+                (lsp--spinner-stop)
+                (lsp--remove-cur-overlays))))
+
+          ;; cleanup session from references to the closed workspace.
```
Comment (MaskRay) [inline lsp-mode.el]: We can capitalize comments since we are updating them.

**Comment by MaskRay** [comment]
Thanks for the patch. I confirm it addresses #1160. Yesterday while I was debugging, I noticed that `lsp--send-no-wait` also checks process-status. For a language server that is stopped after it has run for a while, does `lsp--send-no-wait` or somewhere else need a similar update?

File: lsp-mode.el (hunk 1 above)
Comment (yyoncho) [reply lsp-mode.el]: AFAIK there is early return using cl-lib, I refactored the code to use closure.

File: lsp-mode.el (hunk 2 above)
Comment (yyoncho) [reply lsp-mode.el]: Done, but I think we do not follow this style guide.

**Comment by yyoncho** [comment]
> I noticed that lsp--send-no-wait also checks process-status. For a language server that is stopped after it has run for a while, does lsp--send-no-wait or somewhere else need a similar update?

I changed the code to use conditional-case instead of checking the status of the process upfront. It this should have some positive effect on performance and it will handle your concern.

File: lsp-mode.el [hunk 3]
```diff
@@ -1536,33 +1536,33 @@ WORKSPACE is the workspace that contains the diagnostics."
       "Report new diagnostics to flymake."
       (funcall lsp--flymake-report-fn
                (-some->> (lsp-diagnostics)
-                 (gethash buffer-file-name)
```
Comment (seagle0128) [inline lsp-mode.el]: Bad formats here 😄

Comment (yyoncho) [reply lsp-mode.el]: They are correct but it is GH rendering the indentation in an odd way. Anyway, I removed the irrelevant changes.

**Pull request merged**